1_IGT_PP/Code/R/1_Analyses/plausible_correlations.R

```diff
@@ -0,0 +1,153 @@
+    result <- (1-rho^2)^((n-1)/2)*hyper.term
+    return(result)
+  }
+  
+  .bFunction <- function(n, r, rho) {
+    hyper.term <- Re(hypergeo::genhypergeo(U=c(n/2, n/2), L=(3/2), z=(r*rho)^2))
+    log.term <- 2*(lgamma(n/2)-lgamma((n-1)/2))+((n-1)/2)*log(1-rho^2)
+    result <- 2*r*rho*exp(log.term)*hyper.term
+    return(result)
+  }
+  
+  .hFunction <- function(n, r, rho) {
+    result <- .aFunction(n, r, rho) + .bFunction(n, r, rho)
+    return(result)
+  }
+  
+  .scaledBeta <- function(rho, alpha, beta){
+    result <- 1/2*dbeta((rho+1)/2, alpha, beta)
+    return(result)
+  }
+  
+  .priorRho <- function(rho, kappa=1) {
+    .scaledBeta(rho, 1/kappa, 1/kappa)
+  }
+  
+  fisherZ <- function(r) log((1+r)/(1-r))/2
+  
+  inv.fisherZ <- function(z) {K <- exp(2*z); (K-1)/(K+1)}
+  
+  
+  # Main body
+  
+  # Values spaced around mode
+  qs <- qlogis(seq(0,1,length.out=npoints+2)[-c(1,npoints+2)])
+  rho <- c(-1,inv.fisherZ(fisherZ(r)+qs/sqrt(n)),1)
+  # Get heights
+  if (!is.na(r) && !r==0) {
+    d <- .bf10Exact(n, r, kappa)*.hFunction(n, r, rho)*.priorRho(rho, kappa)
+  } else if (!is.na(r) && r==0) {
+    d <- .bf10JeffreysIntegrate(n, r, kappa)*
+      .jeffreysApproxH(n, r, rho)*.priorRho(rho, kappa)
+  } else return(NA)
+  # Unnormalized approximation funciton for density
+  approxfun(rho,d)
+}
+
+# Estimate population correlation using analytical posterior as described in 
+# Ly, Marsman, & Wagenmakers, 2017, Statistica Neerlandica
+postRav <- function(r, n, spacing=.01, kappa=1,npoints=100,save=FALSE)
+  # r is a vector, returns average density. Can also save unnormalized pdfs
+{
+  funs <- sapply(r,posteriorRho,n=n,npoints=npoints,kappa=kappa)
+  rho <- seq(-1,1,spacing)
+  result <- apply(matrix(unlist(lapply(funs,function(x){
+    out <- x(rho); out/sum(out)
+  })),nrow=length(rho)),1,mean)
+  names(result) <- seq(-1,1,spacing)
+  attr(result,"n") <- n
+  attr(result,"kappa") <- kappa
+  if (save) attr(result,"updfs") <- funs
+  result
+}
+postRav.Density <- function(result)
+  # Produces density class object
+{
+  x.vals <- as.numeric(names(result))
+  result <- result/(diff(range(x.vals))/length(x.vals))
+  out <- list(x=x.vals,y=result,has.na=FALSE,
+              data.name="postRav",call=call("postRav"),
+              bw=mean(diff(x.vals)),n=attr(result,"n")) 
+  class(out) <- "density"
+  out
+}
+
+# Compute the posterior density for population
+pop_correct_MCMC <- function(post_samples, n_subj, kappa) {
+  # Subtract because function fails if posterior samples are too close to 1
+  dens.r <- postRav(n=n_subj, r=post_samples-.1,spacing=.01,kappa=kappa, npoints = 100) 
+  # correct for the subtraction above (shift the density over 10 elements/
+  # .1 in correlation space)
+  dens.r[11:201] <- dens.r[1:191]
+  
+  # Sample from the population-corrected PDF
+  corrected_samples <- approx(
+    cumsum(dens.r)/sum(dens.r),
+    seq(-1,1, length.out=length(dens.r)),
+    runif(10000)
+    )
+  return(corrected_samples$y)
+}
```

Comment:
```suggestion
# Compute the posterior density for population
pop_correct_MCMC <- function(post_samples, n_subj, kappa) {
  # Scale because the method fails when values are close to bounds -1 or 1
  dens.r <- postRav(n=n_subj, r=post_samples*.97,spacing=.01,kappa=kappa, npoints = 100)
  
  # Sample from the population-corrected PDF
  corrected_samples <- approx(
    cumsum(dens.r)/sum(dens.r),
    seq(-1,1, length.out=length(dens.r))*1.06, # rescale to corect scaling
    runif(10000)
    )
  # set anything that goes out of bounds to bound
  corrected_samples$y[corrected_samples$y < -1] <- -1
  corrected_samples$y[corrected_samples$y > 1] <- 1
  return(corrected_samples$y)
}
```

This should work now! Like I noted above, below is some example code to show that test-retest correlations for `betaF` and `betaP` are already estimated on the scale we want them on. In this case,  `pop_correct_MCMC` should return back essentially the same result as the correlations extracted from the model. (note that this code is mean to work with local variables that are available where I made my comment about fixing this above):

```
betaF_from_model = rstan::extract(orl_fit)$betaF[,1,2]
betaF_pop_correct = pop_correct_MCMC(
    iterative_correlations$correlation[iterative_correlations$parameter=="betaF"], 
    49, 1
)

plot(density(betaF_from_model), col="black", xlim=c(-1,1), ylim=c(0,3)) 
par(new=T) 
plot(density(betaF_pop_correct), col="red", xlim=c(-1,1), ylim=c(0,3))
```